File: CellsToServers/HierarchicalTilingApp/Transformation/Transformator.cs
using HierarchicalTilingApp.SumOfSquares;
using System;
using System.Collections.Generic;

namespace HierarchicalTilingApp.Transformation
{
    public class Transformator
    {
        private ShellBuilder shellBuilder;

        public Transformator(ShellBuilder shellBuilder)
        {
            this.shellBuilder = shellBuilder;
        }

        public void transformCellIdxToIndicesArray(int histogramResolution, int[] indicesArray, int cellIdx)
        {
            indicesArray[0] =
                cellIdx / (int)Math.Pow(histogramResolution, indicesArray.Length - 1);
            for (int coordIdx = 1; coordIdx < indicesArray.Length; coordIdx++)
            {
                indicesArray[coordIdx] = cellIdx;
                for (int subCoordIdx = coordIdx - 1; subCoordIdx >= 0; subCoordIdx--)
                {
                    indicesArray[coordIdx] -=
                        (int)Math.Pow(histogramResolution, indicesArray.Length - (subCoordIdx + 1))
                        * indicesArray[subCoordIdx];
                }
                indicesArray[coordIdx] =
                    indicesArray[coordIdx] /
                    (int)Math.Pow(histogramResolution, indicesArray.Length - (coordIdx + 1));
            }
        }
        
        public int[] mergeIndicesArrays(int spaceDimension, int[] outerIndicesArray, int[] innerIndicesArray)
        {
            int[] mergedArrayIndices = new int[2 * spaceDimension];
            for (int idx = 0; idx < spaceDimension; idx++)
            {
                mergedArrayIndices[2 * idx] = outerIndicesArray[idx];
                mergedArrayIndices[2 * idx + 1] = innerIndicesArray[idx];
            }
            return mergedArrayIndices;
        }

        public int[] mergeIndicesArrays(int spaceDimension, int splitNO, int[] outerIndicesArray, 
            int[] innerIndicesArray)
        {
            int[] mergedArrayIndices = new int[2 * spaceDimension + 1];
            mergedArrayIndices[0] = splitNO;
            for (int idx = 0; idx < spaceDimension; idx++)
            {
                mergedArrayIndices[2 * idx + 1] = outerIndicesArray[idx];
                mergedArrayIndices[2 * idx + 2] = innerIndicesArray[idx];
            }
            return mergedArrayIndices;
        }
        
        public bool validateRegionHasEnoughBins(int spaceDimension, int[] indicesArray, int splitNO)
        {
            bool hasEnoughBins = true;
            int binNOInThisRegion = 1;
            for (int idx = 0; idx < spaceDimension; idx++)
            {
                int lowerBound = indicesArray[2 * idx];
                int upperBound = indicesArray[2 * idx + 1];
                binNOInThisRegion *= (upperBound - lowerBound + 1);
            }
            if (binNOInThisRegion <= splitNO)
            {
                hasEnoughBins = false;
            }
            return hasEnoughBins;
        }

        public void splitIndicesArrays(int spaceDimension, int splitDimIdx, int[] indicesArray,
            int componentInSplitDim, out int[] firstPartIndicesArray, out int[] secondPartIndicesArray)
        {
            firstPartIndicesArray = new int[2 * spaceDimension];
            secondPartIndicesArray = new int[2 * spaceDimension];
            for (int idx = 0; idx < spaceDimension; idx++)
            {
                int lowerBound = indicesArray[2 * idx];
                int upperBound = indicesArray[2 * idx + 1];
                firstPartIndicesArray[2 * idx] = lowerBound;
                secondPartIndicesArray[2 * idx + 1] = upperBound;
                if (idx == splitDimIdx)
                {
                    firstPartIndicesArray[2 * idx + 1] = componentInSplitDim;
                    secondPartIndicesArray[2 * idx] = componentInSplitDim + 1;
                }
                else
                {
                    firstPartIndicesArray[2 * idx + 1] = upperBound;
                    secondPartIndicesArray[2 * idx] = lowerBound;
                }
            }
        }
               
        public void initializeObjectiveValueArray(double initializationValue, Array objectiveValueArray)
        {
            for (int[] movingIndicesArray = determineFirstIndicesArray(objectiveValueArray); 
                    movingIndicesArray != null;
                    movingIndicesArray = determineNextIndicesArray(objectiveValueArray, movingIndicesArray))
            {
                objectiveValueArray.SetValue(initializationValue, movingIndicesArray);
            }
        }

        /// <summary>
        /// The following implementation is based on 
        /// stackoverflow.com/questions/9914230/iterate-through-an-array-of-arbitrary-dimension/9914326#9914326
        /// </summary>
        /// <param name="array"></param>
        /// <returns></returns>
        public int[] determineFirstIndicesArray(Array array)
        {
            int spaceDimension = array.Rank;
            int[] firstIndicesArray = new int[spaceDimension];
            for (int dimIdx = 0; dimIdx < spaceDimension; dimIdx++)
            {
                firstIndicesArray[dimIdx] = array.GetLowerBound(dimIdx);
            }
            return firstIndicesArray;
        }

        /// <summary>
        /// The following implementation is based on 
        /// stackoverflow.com/questions/9914230/iterate-through-an-array-of-arbitrary-dimension/9914326#9914326
        /// </summary>
        /// <param name="array"></param>
        /// <param name="previousIndicesArray"></param>
        /// <returns></returns>
        public int[] determineNextIndicesArray(Array array, int[] previousIndicesArray)
        {
            int spaceDimension = array.Rank;
            int[] nextIndicesArray = new int[spaceDimension];
            previousIndicesArray.CopyTo(nextIndicesArray, 0);
            for (int dimIdx = spaceDimension - 1; dimIdx >= 0; --dimIdx)
            {
                nextIndicesArray[dimIdx]++;
                if (nextIndicesArray[dimIdx] <= array.GetUpperBound(dimIdx))
                    return nextIndicesArray;
                nextIndicesArray[dimIdx] = array.GetLowerBound(dimIdx);
            }
            return null;
        }

        public int[] determineIndicesArray(int spaceDimension, int[] extendedIndicesArray)
        {
            int[] indicesArray = new int[2 * spaceDimension];
            for (int idx = 0; idx < spaceDimension; idx++)
            {
                indicesArray[2 * idx] = extendedIndicesArray[2 * idx + 1];
                indicesArray[2 * idx + 1] = extendedIndicesArray[2 * idx + 2];
            }
            return indicesArray;
        }
        
        public Dictionary<int, List<int[]>> convertIntPairsOfShellsToListOfIdxArrays(int histogramResolution, 
            int[] inputIndicesArray, Shell[] shells)
        {
            Dictionary<int, List<int[]>> result = new Dictionary<int, List<int[]>>();
            int shellIdx = 0;
            foreach (var shell in shells)
            {
                List<int[]> indicesArraysInCurrentShell = new List<int[]>();
                IntTuple[] intTuples = shell.getIntTuples();
                foreach (var intTuple in intTuples)
                {
                    int[] currentIndicesArray;
                    if(intTuple.determineIdxArrayRelativeTo(histogramResolution, inputIndicesArray, 
                        out currentIndicesArray))
                    {
                        indicesArraysInCurrentShell.Add(currentIndicesArray);
                    }
                }
                result.Add(shellIdx, indicesArraysInCurrentShell);
                shellIdx++;
            }
            return result;
        }

        public int[] determineFirstIndicesArray(int[] indicesArrayOfRegion)
        {
            int spaceDimension = indicesArrayOfRegion.Length/2;
            int[] firstIndicesArray = new int[spaceDimension];
            for (int dimIdx = 0; dimIdx < spaceDimension; dimIdx++)
            {
                firstIndicesArray[dimIdx] = indicesArrayOfRegion[2 * dimIdx];
            }
            return firstIndicesArray;
        }

        public int[] determineNextIndicesArray(int[] indicesArrayOfRegion, int[] previousIndicesArray)
        {
            int spaceDimension = indicesArrayOfRegion.Length / 2;
            int[] nextIndicesArray = new int[spaceDimension];
            previousIndicesArray.CopyTo(nextIndicesArray, 0);
            for (int dimIdx = spaceDimension - 1; dimIdx >= 0; --dimIdx)
            {
                nextIndicesArray[dimIdx]++;
                int lowerBoundForCurrentDim = indicesArrayOfRegion[2 * dimIdx];
                int upperBoundForCurrentDim = indicesArrayOfRegion[2 * dimIdx + 1];
                if (nextIndicesArray[dimIdx] <= upperBoundForCurrentDim)
                    return nextIndicesArray;
                nextIndicesArray[dimIdx] = lowerBoundForCurrentDim;
            }
            return null;
        }
    }
}

File: CellsToServersCpp/HierarchicalTilingAppCpp/Transformation/Transformator.cpp
#include "Transformator.hpp"

namespace Transformation
{
	int Transformator::calculateCellIdx(int arrayRank, int arrayResolution, int * indicesArray)
	{
		int cellIdx = 0;
		for (int dimIdx = 0; dimIdx < arrayRank; dimIdx++)
		{
			cellIdx += indicesArray[dimIdx] * (int)pow((double)arrayResolution, dimIdx);
		}
		return cellIdx;
	}

	int Transformator::calculateExtendedCellIdx(int arrayRank, int arrayResolution,
		int * extendedIndicesArray)
	{
		int extendedCellIdx = extendedIndicesArray[0];
		for (int dimIdx = 1; dimIdx < arrayRank; dimIdx++)
		{
			extendedCellIdx += extendedIndicesArray[dimIdx] * (int)pow((double)arrayResolution, dimIdx);
		}
		return extendedCellIdx;
	}

	int * Transformator::copyIndicesArray(int spaceDimension, int * inputIndicesArray)
	{
		int * result = new int[spaceDimension];
		for (int idx = 0; idx < spaceDimension; idx++)
		{
			result[idx] = inputIndicesArray[idx];
		}
		return result;
	}

	int * Transformator::extendIndicesArray(int spaceDimension, int * inputIndicesArray)
	{
		int * result = new int[2 * spaceDimension + 1];
		for (int idx = 0; idx < 2 * spaceDimension; idx++)
		{
			result[idx + 1] = inputIndicesArray[idx];
		}
		return result;
	}

	int * Transformator::determineNextIndicesArray(int spaceDimension, int histogramResolution,
		int * previousIndicesArray)
	{
		int * nextIndicesArray = copyIndicesArray(spaceDimension, previousIndicesArray);
		for (int dimIdx = 0; dimIdx < spaceDimension; dimIdx++)
		{
			nextIndicesArray[dimIdx]++;
			if (nextIndicesArray[dimIdx] <= histogramResolution)
				return nextIndicesArray;
			nextIndicesArray[dimIdx] = 0;
		}
		return nullptr;
	}

	int * Transformator::determineNextIndicesArray(int spaceDimension, int histogramResolution,
		int * lowerBoundArray, int * previousIndicesArray)
	{
		int * nextIndicesArray = copyIndicesArray(spaceDimension, previousIndicesArray);
		for (int dimIdx = 0; dimIdx < spaceDimension; dimIdx++)
		{
			nextIndicesArray[dimIdx]++;
			if (nextIndicesArray[dimIdx] <= histogramResolution)
				return nextIndicesArray;
			nextIndicesArray[dimIdx] = lowerBoundArray[dimIdx];
		}
		return nullptr;
	}

	int * Transformator::determineFirstContainedIndicesArray(int spaceDimension,
		int * indicesArrayOfRegion)
	{
		int * firstIndicesArray = new int[spaceDimension];
		for (int dimIdx = 0; dimIdx < spaceDimension; dimIdx++)
		{
			firstIndicesArray[dimIdx] = indicesArrayOfRegion[2 * dimIdx];
		}
		return firstIndicesArray;
	}

	int * Transformator::determineNextContainedIndicesArray(int spaceDimension,
		int * indicesArrayOfRegion,	int * previousIndicesArray)
	{
		int * nextIndicesArray = copyIndicesArray(spaceDimension, previousIndicesArray);
		for (int dimIdx = 0; dimIdx < spaceDimension; dimIdx++)
		{
			nextIndicesArray[dimIdx]++;
			int lowerBoundForCurrentDim = indicesArrayOfRegion[2 * dimIdx];
			int upperBoundForCurrentDim = indicesArrayOfRegion[2 * dimIdx + 1];
			if (nextIndicesArray[dimIdx] <= upperBoundForCurrentDim)
				return nextIndicesArray;
			nextIndicesArray[dimIdx] = lowerBoundForCurrentDim;
		}
		return nullptr;
	}

	int * Transformator::mergeIndicesArrays(int spaceDimension, int * outerIndicesArray,
		int * innerIndicesArray)
	{
		int * mergedArrayIndices = new int[2 * spaceDimension];
		for (int dimIdx = 0; dimIdx < spaceDimension; dimIdx++)
		{
			mergedArrayIndices[2 * dimIdx] = outerIndicesArray[dimIdx];
			mergedArrayIndices[2 * dimIdx + 1] = innerIndicesArray[dimIdx];
		}
		return mergedArrayIndices;
	}

	int * Transformator::mergeIndicesArrays(int spaceDimension, int splitNO, int * outerIndicesArray,
			int * innerIndicesArray)
	{
		int * mergedArrayIndices = new int[2 * spaceDimension + 1];
		mergedArrayIndices[0] = splitNO;
		for (int dimIdx = 0; dimIdx < spaceDimension; dimIdx++)
		{
			mergedArrayIndices[2 * dimIdx + 1] = outerIndicesArray[dimIdx];
			mergedArrayIndices[2 * dimIdx + 2] = innerIndicesArray[dimIdx];
		}
		return mergedArrayIndices;
	}

	int * Transformator::determineIndicesArray(int spaceDimension, int * extendedIndicesArray)
	{
		int * indicesArray = new int[2 * spaceDimension];
		for (int idx = 0; idx < spaceDimension; idx++)
		{
			indicesArray[2 * idx] = extendedIndicesArray[2 * idx + 1];
			indicesArray[2 * idx + 1] = extendedIndicesArray[2 * idx + 2];
		}
		return indicesArray;
	}

	Dictionary_s * Transformator::convertIntPairsOfShellsToListOfIdxArrays(int histogramResolution,
			int * inputIndicesArray, Vector_s shells)
	{
		Dictionary_s * result = new Dictionary_s();
		int shellIdx = 0;
		for (Vector_s::iterator itr = shells.begin(); itr != shells.end(); itr++)
		{
			Shell * shell = *itr;
			Shell_idxs * indicesArraysInCurrentShell = new Shell_idxs();
			Vector_t intTuples = shell->getIntTuples();
			for (Vector_t::iterator subItr = intTuples.begin(); subItr != intTuples.end(); subItr++)
			{
				IntTuple * intTuple = *subItr;
				int * currentIndicesArray = nullptr;
				if(intTuple->determineIdxArrayRelativeTo(histogramResolution, inputIndicesArray,
					currentIndicesArray))
				{
					indicesArraysInCurrentShell->push_back(currentIndicesArray);
				}
			}
			result->insert( std::pair<int, Shell_idxs *>(shellIdx, indicesArraysInCurrentShell) );
			shellIdx++;
		}
		return result;
	}

	bool Transformator::validateRegionHasEnoughBins(int spaceDimension, int * indicesArray, int splitNO)
	{
		bool hasEnoughBins = true;
		int binNOInThisRegion = 1;
		for (int idx = 0; idx < spaceDimension; idx++)
		{
			int lowerBound = indicesArray[2 * idx];
			int upperBound = indicesArray[2 * idx + 1];
			binNOInThisRegion *= (upperBound - lowerBound + 1);
		}
		if (binNOInThisRegion <= splitNO)
		{
			hasEnoughBins = false;
		}
		return hasEnoughBins;
	}

	void Transformator::splitIndicesArrays(int spaceDimension, int splitDimIdx, int * indicesArray,
		int componentInSplitDim, int *& firstPartIndicesArray, int *& secondPartIndicesArray)
	{
		firstPartIndicesArray = new int[2 * spaceDimension];
		secondPartIndicesArray = new int[2 * spaceDimension];
		for (int idx = 0; idx < spaceDimension; idx++)
		{
			int lowerBound = indicesArray[2 * idx];
			int upperBound = indicesArray[2 * idx + 1];
			firstPartIndicesArray[2 * idx] = lowerBound;
			secondPartIndicesArray[2 * idx + 1] = upperBound;
			if (idx == splitDimIdx)
			{
				firstPartIndicesArray[2 * idx + 1] = componentInSplitDim;
				secondPartIndicesArray[2 * idx] = componentInSplitDim + 1;
			}
			else
			{
				firstPartIndicesArray[2 * idx + 1] = upperBound;
				secondPartIndicesArray[2 * idx] = lowerBound;
			}
		}
	}

	int Transformator::determineMaxRange(int spaceDimension, int histogramResolution)
	{
		double temp = pow((double)histogramResolution, (double)spaceDimension);
		temp /= (factorial(spaceDimension) * doubleFactorial(spaceDimension));
		temp *= pow(M_PI * spaceDimension / 2.0, spaceDimension / 2.0);
		int result = (int)ceil(temp);
		return result;
	}

	int Transformator::factorial(int input)
    { 
        int result = 1;
        for (int idx = 1; idx <= input; idx++)
        {
           result *= idx;
        }
        return result;
    }

	int Transformator::doubleFactorial(int input)
    {
       int result;
       if ((input % 2) == 0)
       {
           result = 1;
           for (int idx = 1; idx <= input/2; idx++)
           {
               result *= 2 * idx;
           }
       }
       else
       {
           result = 1;
           for (int idx = 1; idx <= (input + 1) / 2; idx++)
           {
               result *= (2 * idx - 1);
           }
       }
       return result;
   }
}